File: agenttel-node/src/events/emitter.ts
/** Structured event emitter for observability events. */
export type EventHandler = (eventName: string, attributes: Record<string, unknown>) => void;

export class AgentTelEventEmitter {
  private readonly handlers: EventHandler[] = [];

  onEvent(handler: EventHandler): void {
    this.handlers.push(handler);
  }

  emit(eventName: string, attributes: Record<string, unknown> = {}): void {
    for (const handler of this.handlers) {
      try {
        handler(eventName, attributes);
      } catch {
        // handlers should not throw
      }
    }
  }

  emitJSON(eventName: string, data: unknown): void {
    this.emit(eventName, { payload: JSON.stringify(data) });
  }
}

File: agenttel-go/events/emitter.go
// Package events provides structured event emission via OTel-compatible logging.
package events

import (
	"encoding/json"
	"log"
	"sync"
)

// Handler is a callback for receiving emitted events.
type Handler func(eventName string, body map[string]any)

// Emitter emits structured AgentTel events.
type Emitter struct {
	mu       sync.RWMutex
	handlers []Handler
}

// NewEmitter creates an event emitter.
func NewEmitter() *Emitter {
	return &Emitter{}
}

// OnEvent registers a handler for all emitted events.
func (e *Emitter) OnEvent(h Handler) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.handlers = append(e.handlers, h)
}

// Emit sends a structured event to all registered handlers.
func (e *Emitter) Emit(eventName string, body map[string]any) {
	e.mu.RLock()
	defer e.mu.RUnlock()

	for _, h := range e.handlers {
		func() {
			defer func() {
				if r := recover(); r != nil {
					log.Printf("agenttel: event handler panic for %s: %v", eventName, r)
				}
			}()
			h(eventName, body)
		}()
	}
}

// EmitJSON emits an event, serializing the body to JSON for logging.
func (e *Emitter) EmitJSON(eventName string, body map[string]any) string {
	e.Emit(eventName, body)
	data, err := json.Marshal(body)
	if err != nil {
		return "{}"
	}
	return string(data)
}
